src/libsyntax/parse/parser.rs

```diff
@@ -6225,7 +6275,65 @@ impl<'a> Parser<'a> {
             return Ok(Some(macro_def));
         }
 
-        self.parse_macro_use_or_failure(attrs,macros_allowed,attributes_allowed,lo,visibility)
+        // Verify wether we have encountered a struct or method definition where the user forgot to
+        // add the `struct` or `fn` keyword after writing `pub`: `pub S {}`
+        if visibility == Visibility::Public &&
+            self.check_ident() &&
```

Comment:
Isn't it possible these days to have *paths* to macros? e.g., `foo::bar!`?

cc @jseyfried @petrochenkov

Reply:
The way the fall back happens, if the code isn't either `pub foo (` or `pub foo {`, the current macro handling will happen. In the case of `pub foo::bar` this new branch will not be executed. As long as `parse_macro_use_or_failure` does the right thing with paths, this should be fine :)